runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -119,20 +124,42 @@ public Object getPayload(CorfuRuntime runtime) {
     @Override
     public synchronized void releaseBuffer() {
         if (serializedCache != null) {
-            serializedCache.release();
-            if (serializedCache.refCnt() == 0) {
+            serializedCache.buffer.release();
+            if (serializedCache.buffer.refCnt() == 0) {
                 serializedCache = null;
             }
         }
     }
 
     @Override
-    public synchronized void acquireBuffer() {
+    public synchronized void acquireBuffer(boolean metadata) {
         if (serializedCache == null) {
-            serializedCache = Unpooled.buffer();
-            doSerializeInternal(serializedCache);
+            acquireBufferInternal(metadata);
         } else {
-            serializedCache.retain();
+            if (metadata) {
+                serializedCache.buffer.resetReaderIndex();
+                serializedCache.buffer.writerIndex(serializedCache.metadataOffset);
+                doSerializeMetadataInternal(serializedCache.buffer);
+            }
+            serializedCache.buffer.retain();
+        }
+    }
+
+     public synchronized void updateAcquiredBuffer(boolean metadata) {
+        if (serializedCache == null) {
+            return;
```

Comment:
shouldn't we create an instance of serializedCache?

Reply:
This should be called when serializedCache is already acquired

Reply:
If this is only expected to be called if `acquireBuffer` and that would create an instance of `SerializedCache` , then this can never be null, if it is then it should throw an exception.

Reply:
Done.